src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -0,0 +1,223 @@
+        auto it = std::remove_if(pending.begin(), pending.end(),
+            [&] (auto item) -> bool { return item->task->getStorageID() == id; });
+        pending.erase(it, pending.end());
+
+        /// Copy items to wait for their completion
+        std::copy_if(active.begin(), active.end(), std::back_inserter(tasks_to_wait),
+            [&] (auto item) -> bool { return item->task->getStorageID() == id; });
+    }
+
+
+    for (auto & item : tasks_to_wait)
+        item->is_done.wait();
+
+    {
+        std::lock_guard lock(mutex);
+        currently_deleting.erase(id);
+    }
+}
+
+
+void MergeTreeBackgroundExecutor::routine(ItemPtr item)
+{
+    setThreadName(name.c_str());
+
+    auto erase_from_active = [this, item]
+    {
+        active.erase(std::remove(active.begin(), active.end(), item), active.end());
+    };
+
+    try
+    {
+        if (item->task->execute())
+        {
+            std::lock_guard guard(mutex);
+
+            if (currently_deleting.contains(item->task->getStorageID()))
+            {
+                erase_from_active();
+                return;
+            }
+
+            pending.push_back(item);
+            erase_from_active();
+            item->is_done.reset();
+            has_tasks.notify_one();
+            return;
+        }
+
+        std::lock_guard guard(mutex);
+        erase_from_active();
+        has_tasks.notify_one();
+        /// In a situation of a lack of memory this method can throw an exception,
+        /// because it may interact somehow with BackgroundSchedulePool, which may allocate memory
+        /// But it is rather safe, because we have try...catch block here, and another one in ThreadPool.
+        item->task->onCompleted();
+        item->task.reset();
+    }
+    catch (...)
+    {
+        std::lock_guard guard(mutex);
+        erase_from_active();
+        has_tasks.notify_one();
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+        /// Do not want any exceptions
+        try { item->task->onCompleted(); } catch (...) {}
+    }
+}
+
+
+void MergeTreeBackgroundExecutor::schedulerThreadFunction()
+{
+    while (true)
+    {
+        std::unique_lock lock(mutex);
+
+        has_tasks.wait(lock, [this](){ return !pending.empty() || shutdown; });
+
+        if (shutdown)
+            break;
+
+        ItemPtr item = std::move(pending.front());
+        pending.pop_front();
+
+        /// Execute a piece of task
+        bool res = pool.trySchedule([this, item] () mutable
+        {
+            routine(item);
+            /// When storage shutdowns it will wait until all related background tasks
+            /// are finished, because they may want to interact with its fields
+            /// and this will cause segfault.
+            item->is_done.set();
+        });
+
+        if (!res)
+        {
+            active.erase(std::remove(active.begin(), active.end(), item), active.end());
```

Comment:
Why do we remove it from active? This task was not pused yet

Reply:
Bug. Previously pushing to `active` was above this line.